qutip/rcsolve.py

```diff
@@ -0,0 +1,156 @@
+#    (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+#    OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+###############################################################################
+"""
+This module provides exact solvers for a system-bath setup using the
+reaction coordinate method.
+"""
+
+#Author: Neill Lambert, Anubhav Vardhan
+#Contact: nwlambert@gmail.com
+
+__all__ = ['rcsolve']
+
+import warnings
+import numpy as np
+import scipy.sparse as sp
+from numpy import matrix
+from numpy import linalg
+from qutip import spre, spost, sprepost, thermal_dm, mesolve, Odeoptions
+from qutip import tensor, identity, destroy, sigmax, sigmaz, basis, qeye, dims
+
+def rcsolve(Hsys, psi0, tlist, e_ops, Q, wc, alpha, N, w_th, sparse=False,
+            options=None):
+    """
+    Function to solve for an open quantum system using the
+    reaction coordinate (RC) model. 
+
+    Parameters
+    ----------
+    Hsys: Qobj
+        The system hamiltonian.
+    psi0: Qobj
+        Initial state of the system.
+    tlist: List.
+        Time over which system evolves.
+    e_ops: list of :class:`qutip.Qobj` / callback function single
+        Single operator or list of operators for which to evaluate
+        expectation values.
+    Q: Qobj
+        The coupling between system and bath.
+    wc: Float
+        Cutoff frequency.
+    alpha: Float
+        Coupling strength.
+    N: Integer
+        Number of cavity fock states.
+    w_th: Float
+        Temperature. 
+    sparse: Boolean
+        Optional argument to call the sparse eigenstates solver if needed.
+    options : :class:`qutip.Options`
+        With options for the solver.
+     
+    Returns
+    -------
+    output: Result
+        System evolution.
+    """
+    if options is None:
+        options = Options()
+    
+    dot_energy, dot_state = Hsys.eigenstates(sparse=sparse)
+    deltaE = dot_energy[1] - dot_energy[0]
+    if (w_th < deltaE/2):
+        warnings.warn("Given w_th might not provide accurate results")
+    gamma = deltaE / (2 * np.pi * wc)
+    wa = 2 * np.pi * gamma * wc #reaction coordinate frequency
+    g = np.sqrt(np.pi * wa * alpha / 2.0) #reaction coordinate coupling
+    nb = (1 / (np.exp(wa/w_th) -1))
+
+    # Reaction coordinate hamiltonian/operators
+    Nmax = N * 2
+    dimensions = dims(Q)
+    a = tensor(destroy(N), qeye(dimensions[1]))            
+    unit = tensor(qeye(N), qeye(dimensions[1]))
+    Q_exp = tensor(qeye(N), Q)
+    Hsys_exp = tensor(qeye(N),Hsys)
+    e_ops_exp = [tensor(qeye(N), kk) for kk in e_ops]
+
+    na = a.dag() * a    # cavity
+    xa = a.dag() + a
+
+    # decoupled Hamiltonian
+    H0 = wa * a.dag() * a + Hsys_exp
+    #interaction
+    H1 = (g * (a.dag() + a) * Q_exp)
+    H = H0 + H1
+    L = 0
+    PsipreEta = 0
+    PsipreX = 0
+
+    all_energy, all_state = H.eigenstates(sparse=sparse)
+    Apre = spre((a + a.dag()))
+    Apost = spost(a + a.dag())
+    for j in range(Nmax):
```

Comment:
This is going to be slow since you are doing loops over Nmax=N**2.  Could be pushed to Cython?